include/net/mqtt.h

```diff
@@ -0,0 +1,629 @@
+struct mqtt_puback_param {
+	u16_t message_id;
+};
+
+/**@brief Parameters for MQTT publish receive(PUBREC). */
+struct mqtt_pubrec_param {
+	u16_t message_id;
+};
+
+/**@brief Parameters for MQTT publish release(PUBREL). */
+struct mqtt_pubrel_param {
+	u16_t message_id;
+};
+
+/**@brief Parameters for MQTT publish complete(PUBCOMP). */
+struct mqtt_pubcomp_param {
+	u16_t message_id;
+};
+
+/**@brief Parameters for MQTT subscription acknowledgment (SUBACK). */
+struct mqtt_suback_param {
+	u16_t message_id;
+	struct mqtt_binstr return_codes;
+};
+
+/**@brief Parameters for MQTT unsubscribe acknowledgment (UNSUBACK). */
+struct mqtt_unsuback_param {
+	u16_t message_id;
+};
+
+/**@brief Parameters for a publish message. */
+struct mqtt_publish_param {
+	/** Messages including topic, QoS and its payload (if any)
+	 *  to be published.
+	 */
+	struct mqtt_publish_message message;
+
+	/** Message id used for the publish message. Redundant for QoS 0. */
+	u16_t message_id;
+
+	/** Duplicate flag. If 1, it indicates the message is being
+	 *  retransmitted. Has no meaning with QoS 0.
+	 */
+	u8_t dup_flag : 1;
+
+	/** Retain flag. If 1, the message shall be stored persistently
+	 *  by the broker.
+	 */
+	u8_t retain_flag : 1;
+};
+
+/**@brief List of topics in a subscription request. */
+struct mqtt_subscription_list {
+	/** Array containing topics along with QoS for each. */
+	struct mqtt_topic *list;
+
+	/** Number of topics in the subscription list */
+	u32_t list_count;
+
+	/** Message id used to identify subscription request. */
+	u16_t message_id;
+};
+
+/**
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
+ *        Currently, only MQTT_EVT_PUBLISH is accompanied with parameters.
+ */
+union mqtt_evt_param {
+	/** Parameters accompanying MQTT_EVT_CONNACK event. */
+	struct mqtt_connack_param connack;
+
+	/** Parameters accompanying MQTT_EVT_PUBLISH event. */
+	struct mqtt_publish_param publish;
+
+	/** Parameters accompanying MQTT_EVT_PUBACK event. */
+	struct mqtt_puback_param puback;
+
+	/** Parameters accompanying MQTT_EVT_PUBREC event. */
+	struct mqtt_pubrec_param pubrec;
+
+	/** Parameters accompanying MQTT_EVT_PUBREL event. */
+	struct mqtt_pubrel_param pubrel;
+
+	/** Parameters accompanying MQTT_EVT_PUBCOMP event. */
+	struct mqtt_pubcomp_param pubcomp;
+
+	/** Parameters accompanying MQTT_EVT_SUBACK event. */
+	struct mqtt_suback_param suback;
+
+	/** Parameters accompanying MQTT_EVT_UNSUBACK event. */
+	struct mqtt_unsuback_param unsuback;
+};
+
+/**@brief Defined MQTT asynchronous event notified to the application. */
+struct mqtt_evt {
+	/** Identifies the event. */
+	enum mqtt_evt_type type;
+
+	/** Contains parameters (if any) accompanying the event. */
+	union mqtt_evt_param param;
+
+	/** Event result. For example, MQTT_EVT_CONNACK has a result code
+	 *  indicating success or failure code of connection procedure.
+	 */
+	int result;
```

Comment:
It can contain error code with the reason of the failure, updated the description.

Reply:
> It can contain error code with the reason of the failure, updated the description.

Well, I mean that if it's MQTT lib's error, it could fit into u8_t/i8_t. If it's E*, int sounds good to not over-optimize.